File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.cpp
#include <string>
#include <vector>
#include <algorithm>
#define INF 1000000

using namespace std;

int solution(int n, vector<vector<int>> edge) {
    int answer = 0;
        
    vector<int> dist(n+1, INF);
    
    dist[1] = 0;
    for(int i = 1; i <= n; i++){
        for(int j = 0; j < edge.size(); j++){
            int from = edge[j][0];
            int to = edge[j][1];
            
            if(dist[from] != INF && dist[to] > dist[from] + 1) dist[to] = dist[from] + 1;
            else if(dist[to] != INF && dist[from] > dist[to] + 1) dist[from] = dist[to] + 1;
        }
    }
    
    sort(dist.rbegin(), dist.rend() - 1);
    
    int m = -1;
    for(int i = 1; i < n; i++) {
        if(dist[i] != INF && dist[i] >= m){
            m = dist[i];
            answer++;
        }
    }
        
    return answer;
}

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
from collections import deque

def solution(n, edge):
    answer = 0
    dist = [1e9] * n
    dist[0] = 0
    edges = [[] for _ in range(n)]
    for v, u in edge:
        edges[v-1].append(u-1)
        edges[u-1].append(v-1)
        
    move = deque([(0, 0)]) # 거리, 정점
    while move:
        d, v = move.popleft()
        
        for u in edges[v]:
            if dist[u] > dist[v]+1:
                dist[u] = dist[v]+1
                move.append((dist[u], u))
                
    answer = dist.count(max(dist))
    
    return answer